### PR TITLE
Build bootstrap compiler in separate phase

### PR DESCRIPTION
**Pull request by jaredpar**
This moves the infrastructure for building a bootstrap compiler into a separate phase.

The reason for this is to facilitate better performance investigations. Having the bootstrap compiler always build inside our build makes it basically impossible to profile using a freshly built compiler to build our repo. Profiles capture both the build of the bootstrap and the subsequent build. Having separate phases makes it much easier to profile and also test out how changes work as the code outside the bootstrap involves.

This work allows me to now: 

- build the bootstrap compiler
- start the profiler
- rebuild roslyn

Overall I think it's also just a bit cleaner this way.

Further I took the opportunity to clean up the PowerShell code a bit and mostly move us to using `pwsh` instead of having to rely on old PowerShell quirks

### CODE REVIEW

**Comment by jaredpar** [comment]
@dotnet/roslyn-infrastructure PTAL 

Note: the test failures are related to a general Helix outage. This change impacts correctness legs which are all passing so feel it's ready for review.

File: eng/test-build-correctness.ps1 [hunk 1]
```diff
@@ -41,29 +42,35 @@ try {
...
-  Write-Host "$buildBossPath -r `"$RepoRoot/`" -c $configuration -p Roslyn.sln"
-  Exec-Console $buildBossPath "-r `"$RepoRoot/`" -c $configuration" -p Roslyn.sln
+  Exec-Command $buildBossPath "-r `"$RepoRoot/`" -c $configuration -p Roslyn.sln"
   Write-Host ""
 
   # Verify the state of our generated syntax files
   Write-Host "Checking generated compiler files"
   Exec-Script (Join-Path $PSScriptRoot "generate-compiler-code.ps1") "-test -configuration:$configuration"
-  Exec-Console dotnet "tool run dotnet-format whitespace . --folder --include-generated --include src/Compilers/CSharp/Portable/Generated/ src/Compilers/VisualBasic/Portable/Generated/ src/ExpressionEvaluator/VisualBasic/Source/ResultProvider/Generated/ --verify-no-changes"
+  Exec-Command dotnet "tool run dotnet-format whitespace . --folder --include-generated --include src/Compilers/CSharp/Portable/Generated/ src/Compilers/VisualBasic/Portable/Generated/ src/ExpressionEvaluator/VisualBasic/Source/ResultProvider/Generated/ --verify-no-changes"
```
Comment (JoeRobich) [inline eng/test-build-correctness.ps1:65]: Could this be `Exec-DotNet` or does this need to use the dotnet CLI that is on PATH?

**Review by JoeRobich — APPROVED** [review]
I like the switch to pwsh!

File: eng/test-build-correctness.ps1 (hunk 1 above)
Comment (jaredpar) [reply eng/test-build-correctness.ps1:65]: This should be `Exec-DotNet`. This is more working today by accident than design. I will fix that in the follow up PR that I'm just about to push.

**Pull request merged**